File: 제3장/주사위 뒤집기.py
# 반복에 포함되지 않는 수의 개수를 구해보세요.

# 다음 눈을 취득한다
def next_dice(dice):
    top = dice//6**5
    left, right = divmod(dice, 6**(5-top))
    return (right+1)*(6**(top+1)) - (left+1)

count = 0
for i in range(0, 6**6):
    check = []
    # 반복할 때까지 다음 주사위를 찾는다
    while i not in check:
        check.append(i)
        i = next_dice(i)
    # 반복한 위치를 확인하고 반복 대상이 아니라면 카운트
    if check.index(i) != 0:
        count += 1
print(count)

File: 제3장/주사위 뒤집기.js
// 반복에 포함되지 않는 수의 개수를 구해보세요.

function next_dice(dice) {
  let top = parseInt(dice/Math.pow(6, 5))
  let left = parseInt(dice/Math.pow(6, 5-top))
  let right = dice % Math.pow(6, 5-top)
  return (right+1)*Math.pow(6, top+1) - (left+1)
}

let all_dice = new Array(Math.pow(6, 6))
for (i=0; i<Math.pow(6, 6); i++) {
  all_dice[i] = 0
}
for (i=0; i<Math.pow(6, 6); i++) {
  if (all_dice[i] == 0) {
    check = new Array()
    while ((all_dice[i] == 0) && (check.indexOf(i) == -1)) {
      check.push(i)
      i = next_dice(i)
    }
    index = check.indexOf(i)
    if (index >= 0) {
      for (j=0; j<check.length; j++) {
        if (j < index) {
          all_dice[check[j]] = 1
        }
        else {
          all_dice[check[j]] = 2
        }
      }
    }
    else {
      for (j=0; j<check.length; j++) {
        all_dice[check[j]] = 1
      }
    }
  }
}
cnt = 0
for (i=0; i<Math.pow(6, 6); i++) {
  if (all_dice[i] == 1) {
    cnt++
  }
}
console.log(cnt)